r/R/dplyr.R

```diff
@@ -0,0 +1,92 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+select.RecordBatch <- function(.data, ...) {
+  .data$selected_columns <- c(.data$selected_columns, list(quos(...)))
+  .data
+}
+
+filter.RecordBatch <- function(.data, ..., .preserve = FALSE) {
+  .data$filtered_rows <- c(.data$filtered_rows, quos(...))
+  .data
+}
+
+collect.RecordBatch <- function(x, ...) {
+  filters <- evaluate_filters(x)
+  colnames <- names(x)
+  for (q in x$selected_columns) {
+    colnames <- vars_select(colnames, !!!q)
```

Comment:
The use of `!!!` suggests that `q` is a list so that `x$selected_column` is a list of lists?

Reply:
i.e. I don't _think_ you need the extra `list()` in the `select.RecordBatch()` method; that also makes it more consistent with `filter.RecordBatch`

Reply:
It's possible I'm missing a subtlety here, but I think there's a relevant difference in behavior between `select` and `filter`. For `filter`, I can flatten the list because you evaluate filters in sequence. I.e. 

```r
df %>%
  filter(var1 == 2, var3 == "d") %>%
  filter(var4 > 12)

df %>%
  filter(var1 == 2) %>%
  filter(var3 == "d") %>%
  filter(var4 > 12)

df %>%
  filter(var1 == 2, var3 == "d", var4 > 12)
```

all yield the same thing. But

```r
df %>%
  select(var1, var2, var3)

df %>%
  select(var1) %>%
  select(var2) %>%
  select(var3)
```

are not equivalent.

Reply:
Oh I see, that makes sense. But note you can't change the order of `filter()` and `select()` so your current approach is going to fall down anyway. Maybe it's worth having a video chat so I can tell you how I'd attack this problem in general.